androidApp/src/main/java/com/example/tasky/android/alarm/data/AlarmScheduler.kt

```diff
@@ -60,4 +62,18 @@ class AlarmScheduler(
 
         alarmRepository.deleteAgendaAlarm(requestCode = requestCode)
     }
+
+    override fun schedulePeriodicSyncAgenda() {
+        val alarmIntent =
+            Intent(context, SyncAgendaReceiver::class.java).let { intent ->
+                PendingIntent.getBroadcast(context, 0, intent, PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE)
+            }
+
+        alarmManager.setInexactRepeating(
+            AlarmManager.ELAPSED_REALTIME_WAKEUP,
+            SystemClock.elapsedRealtime(),
+            AlarmManager.INTERVAL_HALF_HOUR,
+            alarmIntent,
+        )
+    }
```

Comment:
To guarantee the execution and better manage the work, Android has the `WorkManager` class. An example to fetch all agendas would be the following:

```
class FetchFullAgendaWorker(
  context: Context,
  params: WorkerParameters,
  // provide your own dependencies
  private val agendaRepository: AgendaRepository,
) : CoroutineWorker(context, params) { // <— Supports coroutines
  
  override suspend fun doWork(): Result { // <— Listenable.Result type from WorkManager
    // You can even schedule retries and manage your work based on the amount of retries
    if(runAttemptCount >= 5) {
      return Result.failure()
    }
    
    return when(val result = agendaRepository.fetchFullAgenda()) {
      is com.plcoding.tasky.core.domain.util.Result.Error -> {
        result.error.toWorkerResult() // <— Extension function to make domain DataError to Worker Result
      }
      is com.plcoding.tasky.core.domain.util.Result.Success -> Result.success()
    }
  }

  companion object {
    const val TAG = “fetch_full_agenda_worker”
  }
}
```

You can then build the work request in your Sync Scheduler implementation class as follows:
```
  val workRequest = PeriodicWorkRequestBuilder<FetchFullAgendaWorker>(30.minutes.toJavaDuration())
    .addTag(FetchFullAgendaWorker.TAG)
    .setConstraints(
      Constraints.Builder()
        .setRequiredNetworkType(NetworkType.CONNECTED)
        .build()
    )
    .setBackoffCriteria(
      backoffPolicy = BackoffPolicy.EXPONENTIAL,
      backoffDelay = 2000L,
      timeUnit = TimeUnit.MILLISECONDS
    )
    .setInitialDelay(
      duration = 30,
      timeUnit = TimeUnit.MINUTES
    )
    .build()
```
You can use Koin to inject those custom constructor dependencies:
```
val koinModule = module {
  workerOf(::FetchFullAgendaWorker)
}
```

And here is that extension function 😊
```
fun DataError.toWorkerResult(): ListenableWorker.Result {
  return when(this) {
    DataError.Local.DISK_FULL -> ListenableWorker.Result.failure()
    DataError.Network.REQUEST_TIMEOUT -> ListenableWorker.Result.retry()
    DataError.Network.UNAUTHORIZED -> ListenableWorker.Result.retry()
    DataError.Network.CONFLICT -> ListenableWorker.Result.retry()
    DataError.Network.TOO_MANY_REQUESTS -> ListenableWorker.Result.retry()
    DataError.Network.NO_INTERNET -> ListenableWorker.Result.retry()
    DataError.Network.PAYLOAD_TOO_LARGE -> ListenableWorker.Result.failure()
    DataError.Network.SERVER_ERROR -> ListenableWorker.Result.retry()
    DataError.Network.SERIALIZATION -> ListenableWorker.Result.failure()
    DataError.Network.UNKNOWN -> ListenableWorker.Result.failure()
  }
}
```